File: Comb_Sort/comb_sort.js
/* 
Function to implement Comb sort using JavaScript
Example : If input array is [9,7,8,3,2], then after implementing this code the output array will be [2,3,7,8,9].  
*/

function combSort(arr)
{
    // the shrink factor is taken to be 1.3
    var gap = Math.floor(arr.length / 1.3);
    
    /* loop to reduce gap and implement sorting */ 
    while(gap > 0){
        for(var i=0; i+gap < arr.length; i++){
            // swap the values using a temporary variable
                if(arr[i] > arr[i+gap]){
                    var next = arr[i+gap];
                    arr[i+gap] = arr[i];
                    arr[i] = next;
                 }
        }
        gap=Math.floor(gap / 1.3);
    }
    return arr;
};

/* psuedo data for array */
var arr = [3,9,8,2,5,1];

/* prints the solution */
console.log('The sorted array is' + combSort(arr));

/*
    Output : 1, 2, 3, 5, 8, 9
*/

File: Comb_Sort/Comb_Sort.kt
// Kotlin Code for Comb Sort 

import java.lang.Math;
class CombSort
{ 
    // Function to swap elements at index a and b
    fun swap(int:arr[], int:a, int:b):void
    {
        int temp = arr[a]; 
        arr[a] = arr[b]; 
        arr[b] = temp; 
    }

    // Function to sort arr[] using Comb Sort 
    fun combSort(int:arr[], int:n):void
    { 
        // initialize gap value to array length
        int gap = n;
        int flag = 1;
        while (gap > 1 || flag == 1 ) 
        { 
            // Update gap value by using shrink factor 1.3
            gap = (gap * 10) / 13;
            gap = Math.max(1, gap);
            flag = 0;
            // Compare all elements with the obtained gap value
            for (i in 0 until (n-gap)) 
            { 
                if (arr[i] > arr[i+gap]) 
                {
                    swap(arr, i, i + gap);
                    flag = 1; 
                } 
            } 
        }
    } 

    // Main function
    fun main()
    { 
        CombSort obj = new CombSort(); 
        var read = Scanner(System.`in`)
        println("Enter the size of Array:")
        val arrSize = read.nextLine().toInt()
        var arr = IntArray(arrSize)
        println("Enter elements")
        for(i in 0 until arrSize)
        {
            arr[i] = read.nextLine().toInt()
        }
        int n = arrSize;
        obj.combSort(arr, n);

        println("Array after sorting:");
        for (i in 0 until arrSize) 
        {
            println(arr[i] + " ");
        }
    } 
}

/* 
Input
-10, 50, 20, 0, 15, -25, 30
Output
-25 -10 0 15 20 30 50
*/
